[RDY] vim-patch:8.0.{0643-0646}

Billy4195 opened the pull request:
#### vim-patch:8.0.0644: the timeout for 'hlsearch' is not tested

Problem:    There is no test for 'hlsearch' timing out.
Solution:   Add a test.
https://github.com/vim/vim/commit/5b1affefd0e96154517ec6f71300086ae6d22d24


#### vim-patch:8.0.0645: no error for illegal back reference in NFA engine

Problem:    The new regexp engine does not give an error for using a back
            reference where it is not allowed. (Dominique Pelle)
Solution:   Check the back reference like the old engine. (closes vim/vim#1774)
https://github.com/vim/vim/commit/1ef9bbe215e13a273e74fccaddd8fc5a42c76b6e

Billy4195 committed:
vim-patch:8.0.0644: the timeout for 'hlsearch' is not tested

Problem:    There is no test for 'hlsearch' timing out.
Solution:   Add a test.
https://github.com/vim/vim/commit/5b1affefd0e96154517ec6f71300086ae6d22d24

Billy4195 committed:
vim-patch:8.0.0645: no error for illegal back reference in NFA engine

Problem:    The new regexp engine does not give an error for using a back
            reference where it is not allowed. (Dominique Pelle)
Solution:   Check the back reference like the old engine. (closes vim/vim#1774)
https://github.com/vim/vim/commit/1ef9bbe215e13a273e74fccaddd8fc5a42c76b6e

justinmk commented:
Thanks! A `src/nvim/testdir/...` test is failing now (see Travis; these tests don't run on QuickBuild/AppVeyor):

```
From test_hlsearch.vim:
Found errors in Test_hlsearch_hangs():
function RunTheTest[35]..Test_hlsearch_hangs line 12: Expected True but got 0
```

You can [run that test locally](https://github.com/neovim/neovim/wiki/Building-Neovim#running-tests) like this:

    TEST_FILE=test_hlsearch.res make oldtest

Billy4195 commented:
@justinmk Got it.
I will try to fix it.

justinmk commented:
8.0.0644 only added a test. This is a hint that a previous patch (probably not far behind 8.0.0644, check 8.0.06xx patches) added the actual fix.

Billy4195 committed:
vim-patch:8.0.0646: the hlsearch test fails on fast systems

Problem:    The hlsearch test fails on fast systems.
Solution:   Make the search pattern slower.  Fix that the old regexp engine
            doesn't timeout properly.
https://github.com/vim/vim/commit/0946326580e6f034fe2c88d041407ea0fde980ab

Billy4195 commented:
@justinmk Passed the old test that mentioned above, but I have no idea why the CI failed in following cases:
1. xcode 10.1 c & gcc: It failed when installing dependency [29/58]. 
> Because I don't have any macOS environment, I can't test it locally.
2. lint check: coding style checking
> I have run it locally, it popped out a lot of error msgs, but it seems that the coding style I followed is everywhere, it is weird.

justinmk commented:
> xcode 10.1 c & gcc: It failed when installing dependency [29/58].

CI (build servers) can be unreliable. I restarted that job.

> I have run it locally, it popped out a lot of error msgs, but it seems that the coding style I followed is everywhere, it is weird.

see the note about the linter in 
https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#automated-builds-ci

Billy4195 committed:
lint: fix coding style

Billy4195 commented:
@justinmk How can I fix the coverity problem in CI test? I want to test it locally, but I can't see the result of `codecov/patch`, the `Details` link is failed.

justinmk commented:
@Billy4195 The failing "Details" page is a bug of codecov (note: "coverity" is a different, unrelated service). For Vim patches, don't worry about failed codecov. 

This PR looks good, just mark it RDY and it will be reviewed soon.

janlazo commented:
I tried merging these patches before on https://github.com/neovim/neovim/pull/8662 but 8.0.0643 was failing a functional test. I guess 8.0.0643 is an isolated patch from 8.0.0644-7 and other regexp/syntax patches.

Billy4195 commented:
@janlazo Do you mean I should add 8.0.0643 to this PR as well?

justinmk commented:
@Billy4195 This looks good except as hinted, 8.0.0643 is needed. Because currently this is missing the `timed_out` parameter which was added in https://github.com/vim/vim/commit/0946326580e6f034fe2c88d041407ea0fde980ab . We can't half-merge patches, that would be very confusing.

Billy4195 committed:
vim-patch:8.0.0643: when a pattern search is slow Vim becomes unusable

Problem:    When 'hlsearch' is set and matching with the last search pattern
            is very slow, Vim becomes unusable.  Cannot quit search by
            pressing CTRL-C.
Solution:   When the search times out set a flag and don't try again.  Check
            for timeout and CTRL-C in NFA loop that adds states.
https://github.com/vim/vim/commit/fbd0b0af6800f6ff89857863d6a07ea03f09ff6c

Billy4195 commented:
@justinmk To the best of my understanding, the declaration of `nfa_did_time_out()` will be generated by script, but the CI failed, or do I still need to add something?

justinmk commented:
@Billy4195 I think Travis is failing because of a cache issue, because I merged an update to a library yesterday. I cleared its cache and restarted the build. If that doesn't work, please try rebasing on lastest master.

Billy4195 commented:
@justinmk I found the problem and fixed it

justinmk commented on src/nvim/regexp.c:
```diff
@@ -3814,6 +3821,13 @@ regmatch (
+      if (tm != NULL && ++tm_count == 100) {
+        tm_count = 0;
+        if (profile_passed_limit(*tm)) {
+          status = RA_FAIL;
```
the original patch also has this code:

```
if (timed_out != NULL) {
  *timed_out = TRUE;
}
```

why not include that?

justinmk commented on src/nvim/regexp.c:
```diff
@@ -3768,9 +3771,11 @@ static long bl_maxval;
+regmatch(
+    char_u *scan,               // Current node.
+    proftime_T *tm,
+    int     *timed_out
```
original patch added documentation (comments) to these parameters. Why not include that?

I will fix this in the merge-commit.

justinmk commented on src/nvim/regexp.c:
```diff
@@ -1210,27 +1210,30 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
+/// Return TRUE if the back reference is legal. We must have seen the close
+/// brace.
+/// TODO(billy4195): Should also check that we don't refer to something that is
+///   repeated
```
If you didn't write the TODO, it should be `TODO(vim)`. I'll fix this.

justinmk commented:
@Billy4195 Nice work! This is very helpful, thanks.
